### PR TITLE
[FIX] Throw missing subcommand argument on parse() and adapt misleading tutorial.

### LINKED ISSUE #2172: Subcommand parser throws on constructions
### Description 

See the example: https://docs.seqan.de/seqan/3-master-user/subcommand_arg_parse.html
Source is here https://github.com/seqan/seqan3/blob/master/doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp
If you have set up a directory to build the snippet tests: `make subcommand_arg_parse_snippet`

If the provided subparser is invalid, the constructor will throw, see https://github.com/seqan/seqan3/blob/master/include/seqan3/argument_parser/argument_parser.hpp#L684

The snippet suggests that it should be handled with ifs.
I would expect it to work the way it is in the snippets or at least throw when parsing and not on construction.

Currently, the snippet works like this:
```console
./subcommand_arg_parse_snippet hello
terminate called after throwing an instance of 'seqan3::too_few_arguments'
  what():  Please specify which sub program you want to use (one of [push,pull]). Use -h/--help for more information.
Aborted
```
But it suggests that the output will look like this:
```console
./subcommand_arg_parse_snippet hello
terminate called after throwing an instance of 'std::logic_error'
  what():  I do not know sub parser hello
Aborted
```

### Acceptance Criteria

- [ ] test 1
- [ ] test 2
- [ ] test 3

### Tasks

 

- [ ] task 1
- [ ] task 2
- [ ] task 3

### Definition of Done

- [ ] Implementation and design approved
- [ ] Unit tests pass
- [ ] Test coverage = 100%
- [ ] Microbenchmarks added and/or affected microbenchmarks < 5% performance drop
- [ ] API documentation added
- [ ] Tutorial/teaching material added
- [ ] Test suite compiles in less than 30 seconds (on travis)
- [ ] Changelog entry added

### PR DESCRIPTION
**Pull request by smehringer**
Resolves #2172

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -37,6 +37,14 @@ If possible, provide tooling that performs the changes, e.g. a shell-script.
 
 * The `seqan3::fm_index_cursor` exposes its suffix array interval ([\#2076](https://github.com/seqan/seqan3/pull/2076)).
 
+## Notable Bug-fixes
+
+### Argument Parser
+
+* Argument parsing with subcommands: If the user forgets or misspells the subcommand, the error is thrown when calling
+  `seqan3::argument_parser::parse()` not on constructing of the `seqan3::argument_parser`
```
Comment (eseiler) [inline CHANGELOG.md:45]:
```suggestion
  `seqan3::argument_parser::parse()` and not on construction of the `seqan3::argument_parser`
```

File: doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp [hunk 2]
```diff
@@ -101,9 +100,12 @@ int main(int argc, char const ** argv)
...
     if (sub_parser.info.app_name == std::string_view{"mygit-pull"})
-        run_git_pull(sub_parser);
+        return run_git_pull(sub_parser);
     else if (sub_parser.info.app_name == std::string_view{"mygit-push"})
-        run_git_push(sub_parser);
+        return run_git_push(sub_parser);
     else
-        throw std::logic_error{"I do not know sub parser " + sub_parser.info.app_name};
+        std::cout << "Did you catch the wrong names? The subparser is called: " << sub_parser.info.app_name;
+        // Note: If you set up your if branches correctly, you should never reach this else block and you can omit it.
```
Comment (eseiler) [inline doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp:108]: Why would we intentionally do it wrong? :)

Maybe you can rephrase it to state more clearly that you only should reach this if you add a subparser and forget to handle it down here (otherwise it would've thrown in `parse()`)

**Review by eseiler — CHANGES_REQUESTED** [review]
You have a few failing tests in the argparser

File: doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp (hunk 2 above)
Comment (smehringer) [reply doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp]: Good point, I tried :) Take a look and make a suggestion if it is still confusing.

File: doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp [hunk 3]
```diff
@@ -101,9 +100,12 @@ int main(int argc, char const ** argv)
...
 
     if (sub_parser.info.app_name == std::string_view{"mygit-pull"})
-        run_git_pull(sub_parser);
+        return run_git_pull(sub_parser);
     else if (sub_parser.info.app_name == std::string_view{"mygit-push"})
-        run_git_push(sub_parser);
+        return run_git_push(sub_parser);
     else
-        throw std::logic_error{"I do not know sub parser " + sub_parser.info.app_name};
+        std::cout << "Unhandled subparser named " << sub_parser.info.app_name;
```
Comment (eseiler) [inline doc/howto/subcommand_argument_parser/subcommand_arg_parse.cpp:107]:
```suggestion
        std::cout << "Unhandled subparser named " << sub_parser.info.app_name << '\n';
```

**Review by eseiler — APPROVED** [review]

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 4]
```diff
@@ -183,17 +183,20 @@ class argument_parser
...
     {
         if (!std::regex_match(app_name, app_name_regex))
-            throw design_error{"The application name must only contain alpha-numeric characters "
-                                               "or '_' and '-' (regex: \"^[a-zA-Z0-9_-]+$\")."};
-        for (auto & sub : subcommands)
+            throw design_error{"The application name must only contain alpha-numeric characters or '_' and '-' "
+                               "(regex: \"^[a-zA-Z0-9_-]+$\")."};
+
+        for (auto & sub : this->subcommands)
             if (!std::regex_match(sub, std::regex{"^[a-zA-Z0-9_]+$"}))
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp:194]: What is the difference between the app_name_regex and this one? And why does it have different requirements then the app name? For Example no `-` sign?

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 5]
```diff
@@ -554,11 +564,13 @@ class argument_parser
     //!\brief Stores the sub-parser in case \link subcommand_arg_parse subcommand parsing \endlink is enabled.
     std::unique_ptr<argument_parser> sub_parser{nullptr};
 
+    //!\brief Stores the sub-parser names in case \link subcommand_arg_parse subcommand parsing \endlink is enabled.
+    std::vector<std::string> subcommands;
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp:568]:
```suggestion
    std::vector<std::string> subcommands{};
```

File: test/unit/argument_parser/format_parse_test.cpp [hunk 6]
```diff
@@ -921,9 +921,10 @@ TEST(parse_test, subcommand_argument_parser_success)
 
     // incorrect sub command
     {
```
Comment (rrahn) [inline test/unit/argument_parser/format_parse_test.cpp:923]: please add the issue number here to track that this was a related change.

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 7]
```diff
@@ -409,8 +419,8 @@ class argument_parser
     {
         if (sub_parser == nullptr)
         {
-            throw design_error("You did not enable subcommand parsing on construction "
-                                      "so you cannot access the sub-parser!");
+            throw design_error("You did not enable subcommand parsing on construction so you cannot access the "
+                               "sub-parser!");
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp:423]: The message is misleading. I thought I have to set a bool or something to true.
Maybe: 
```suggestion
            throw design_error("No subcommand was provided at the construction of the argument parser!");
```

**Review by rrahn — CHANGES_REQUESTED** [review]
Looks good. Some minor questions/remarks.

File: test/unit/argument_parser/format_parse_test.cpp (hunk 6 above)
Comment (smehringer) [reply test/unit/argument_parser/format_parse_test.cpp:923]:
```suggestion
    { // see issue https://github.com/seqan/seqan3/issues/2172
```

File: include/seqan3/argument_parser/argument_parser.hpp (hunk 4 above)
Comment (smehringer) [reply include/seqan3/argument_parser/argument_parser.hpp:194]: Let me check..
So the subcommand parser name is automatically build with `appname``-``subcommand`, that's why the `-`. I don't remember adding the restriction but it was most probably me. I currently see no reason why this shouldn't be allowed.
I will fix that in a follow up PR though

**Comment by smehringer** [comment]
@rrahn polite ping

File: include/seqan3/argument_parser/argument_parser.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/argument_parser/argument_parser.hpp:194]: So the design is that the subcommand parser needs to be called `foo-bar` insteaf of `foo bar`?

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 8]
```diff
@@ -391,6 +394,13 @@ class argument_parser
 
         detail::version_checker app_version{info.app_name, info.version, info.url};
 
+        if (std::holds_alternative<detail::format_parse>(format) && !subcommands.empty() && sub_parser == nullptr)
+        {
+            throw too_few_arguments{detail::to_string("You either forgot or misspelled the subcommand! Please specify"
+                                                      " which sub program you want to use: one of ", subcommands,
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp:400]:
```suggestion
                                                      " which sub-program you want to use: one of ", subcommands,
```

**Review by rrahn — APPROVED** [review]
Looks good. Can be rebased. Some minor documentation issue.

File: include/seqan3/argument_parser/argument_parser.hpp (hunk 4 above)
Comment (smehringer) [reply include/seqan3/argument_parser/argument_parser.hpp:194]: no the app is still called with `./foo bar` on the command line. But the app name internally `subparser.info.app_name` is called `foo-bar`. A name is required and for the subparser I set the name internally, I guess I chose to add the top level parser to it :shrug: could disambiguate stuff in the future or with multiple nested parsers.

Comment (smehringer) [reply include/seqan3/argument_parser/argument_parser.hpp:194]: I added a follow up issue to track this https://github.com/seqan/product_backlog/issues/234
